server: add NetworkConnectivity endpoint to get network connection statuses between nodes

koorosh opened the pull request:
This patch introduces new endpoint which aimed to provide information
about network connection status between nodes rather than nodes statuses
(that is done with node liveness api).

`NetworkConnectivity` endpoint relies on gossip client to get all known
nodes in the cluster and then checks `rpcCtx.ConnHealth` for every peer.
It can tell us following:
- connection is live if no error returned;
- `ErrNoConnection` returned if nodes don't have connection;
- other errors indicate that network connection is unhealthy;

This functionality should not care about whether node is decommissioned,
shutdown or whatever.

Later on, it'll be used in conjunction with node liveness statuses to
distinguish more specific reasons why network connection is broken.
For instance,
- if node liveness status is DEAD and connection is failed,
then it should not be considered as network issue.
- if node liveness status is LIVE/DECOMMISSIONING and network
connection is failed, then it is network partitioning;

Release note: None

Part of #58033
Part of #96101

Epic: https://cockroachlabs.atlassian.net/browse/CRDB-21710

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95429)
<!-- Reviewable:end -->

koorosh commented on pkg/gossip/gossip.go:
```diff
@@ -521,6 +521,12 @@ func (g *Gossip) GetAddresses() []util.UnresolvedAddr {
 	return append([]util.UnresolvedAddr(nil), g.addresses...)
 }
 
+func (g *Gossip) GetBootstrapAddresses() map[util.UnresolvedAddr]roachpb.NodeID {
```
`GetBootstrapAddresses` is used to get list of known node address in cluster as a source of known nodes.

koorosh commented on pkg/gossip/gossip.proto:
```diff
@@ -130,7 +130,7 @@ message InfoStatus {
   ClientStatus client = 2 [(gogoproto.nullable) = false];
   ServerStatus server = 3 [(gogoproto.nullable) = false];
-  Connectivity connectivity = 4 [(gogoproto.nullable) = false];
+  Connectivity connectivity = 4 [(gogoproto.nullable) = false, deprecated = true];
```
Not relevant to current work. Relates to following PR: https://github.com/cockroachdb/cockroach/pull/89613

koorosh commented on pkg/server/status.go:
```diff
@@ -2832,6 +2917,90 @@ func (s *statusServer) iterateNodes(
+// iterateNodes iterates nodeFn over all non-removed nodes concurrently.
+// It then calls nodeResponse for every valid result of nodeFn, and
+// nodeError on every error result.
+func (s *statusServer) callNodes(
```
`callNodes` is identical to `dialNodes` function with only difference that nodes to dial are provided as argument instead of getting them from `s.serverIterator.getAllNodes(ctx)` (as commented out below).
It allows to dial specific nodes (known to gossip client in our case).

tbg reviewed (changes requested):
The basic ingredients look good.

Would you mind sending the changes for the `rpc` package (`OnDisconnect()`) out separately including a unit test? We want to make sure that if a connection goes bad, we no longer claim to have a healthy latency for it. On that PR we should also manually verify this fact by actually looking at the matrix under a partition. We might be able to backport this, so it's useful to separate it out.

The basic approach for the network latency endpoint seems fine, but a lot of plumbing is getting introduced that I'm unsure of since I don't have the background. Why are we introducing new protos? What is the current network latency matrix using? Is that going to be deprecated? Why isn't it being deprecated here? Etc.

The network connectivity code (that I've just glanced over, not reviewed) could could be architected for better testability. Ideally you have a method that receives only the nodeID->addr mapping and a latencies map and constructs the result, and that is being unit tested. All of the server glue can come separately (in separate commits/PRs) and these will be "dumb" PRs that don't do anything too interesting.

It looks like you have some trouble using `dialNodes` in the context you're in (leading t `callNodes` being introduced). May I suggest a separate (separate PR is fine and easier to handle, I'll also take a separate commit) PR where you adjust `dialNodes` to work for your use case as well.

Santamaura reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @erikgrinaker, @koorosh, and @nvanbenschoten)

___
*[`pkg/server/serverpb/status.proto` line 1898 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NN89iqm3BZj_1_wlt_h:-NN89iqm3BZj_1_wlt_i:b9cn5sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f9ebf3537da999f1ed756afd9a09e7e74146caf/pkg/server/serverpb/status.proto#L1898)):*
> ```protobuf
> }
> 
> message NetworkConnectivityResponse {
> ```

I think in general the structure of the endpoint has almost everything we need, but one question I have is how will the ui be aware about the edge case we talked about where a node has not attempted to connect with another node? To me it seems like there is no way currently to differentiate between not being able to connect with another node because of issues and not being able to connect with another node because it has not been attempted. 


<!-- Sent from Reviewable.io -->

Santamaura reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @erikgrinaker, @koorosh, and @nvanbenschoten)

___
*[`pkg/server/serverpb/status.proto` line 1898 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NN89iqm3BZj_1_wlt_h:-NNbUgawCl-IkdAOuyDP:b-intwc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f9ebf3537da999f1ed756afd9a09e7e74146caf/pkg/server/serverpb/status.proto#L1898)):*
<details><summary><i>Previously, Santamaura (Alex Santamaura) wrote…</i></summary><blockquote>

I think in general the structure of the endpoint has almost everything we need, but one question I have is how will the ui be aware about the edge case we talked about where a node has not attempted to connect with another node? To me it seems like there is no way currently to differentiate between not being able to connect with another node because of issues and not being able to connect with another node because it has not been attempted. 
</blockquote></details>

Actually is there a reason we need to return 2 maps? Could we return only 1 map and instead just add on the liveness state e.g. something like:



_Code snippet:_
```protobuf
message NetworkConnectivityResponse {
  message NodeConnectivity {
    map<int32, int64> latencies = 1 [
      (gogoproto.nullable) = false,
      (gogoproto.castkey) = "github.com/cockroachdb/cockroach/pkg/roachpb.NodeID"
    ];
    kv.kvserver.liveness.livenesspb.Liveness liveness = 2
  };
  map<int32, NodeConnectivity> connectivity_by_node_id = 1 [
    (gogoproto.nullable) = false,
    (gogoproto.castkey) = "github.com/cockroachdb/cockroach/pkg/roachpb.NodeID"
  ];
}
```


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @erikgrinaker, @nvanbenschoten, and @Santamaura)

___
*[`pkg/server/serverpb/status.proto` line 1898 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NN89iqm3BZj_1_wlt_h:-NNlAP9HD8zWBLj8wC7k:b37ddix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f9ebf3537da999f1ed756afd9a09e7e74146caf/pkg/server/serverpb/status.proto#L1898)):*
<details><summary><i>Previously, Santamaura (Alex Santamaura) wrote…</i></summary><blockquote>

Actually is there a reason we need to return 2 maps? Could we return only 1 map and instead just add on the liveness state e.g. something like:
</blockquote></details>

That makes sense, it can be grouped all together. Do we still need `NodeConnectivity` info here?


<!-- Sent from Reviewable.io -->

Santamaura commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @erikgrinaker, @nvanbenschoten, and @Santamaura)
> 
> _[`pkg/server/serverpb/status.proto` line 1898 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NN89iqm3BZj_1_wlt_h:-NNlAP9HD8zWBLj8wC7k:b37ddix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f9ebf3537da999f1ed756afd9a09e7e74146caf/pkg/server/serverpb/status.proto#L1898)):_
> 
> Previously, Santamaura (Alex Santamaura) wrote…
> That makes sense, it can be grouped all together. Do we still need `NodeConnectivity` info here?

I think so, we want it to hold both the liveness state and latency map so when the payload is received on the FE i'm imagining something like `{
  1: {
  latencies: {2: 100, 3:200},
  liveness: 0,
},
...etc
}`
Let me know if it doesn't make sense.

koorosh committed:
status, server: expose network connectivity endpoint

This patch introduces new endpoint which aimed to provide information
about network connection status between nodes rather than nodes statuses
(that is done with node liveness api).

`NetworkConnectivity` endpoint relies on gossip client to get all known
nodes in the cluster and then checks `rpcCtx.ConnHealth` for every peer.
It can tell us following:
- connection is live if no error returned;
- `ErrNoConnection` returned if node is attempting to connect
to target node;
- other errors indicate that network connection is unhealthy;

This functionality should not care about whether node is decommissioned,
shutdown or whatever.

Later on, it'll be used in conjunction with node liveness statuses to
distinguish more specific reasons why network connection is broken.
For instance,
- if node liveness status is DEAD and connection is failed,
then it should not be considered as network issue.
- if node liveness status is LIVE/DECOMMISSIONING and network
connection is failed, then it is network partitioning;

Release note: None

andrewbaptist commented:
I wanted to check what the status of this was. I think this is generally a good change, and should be able to be finished and merged now.

koorosh commented:
> I wanted to check what the status of this was. I think this is generally a good change, and should be able to be finished and merged now.

cc @andrewbaptist , I've updated PR in a way to collect both latencies and connection states based on whether `ConnHealth` returns error or not.

It requires follow up work to ensure that `ConnHealth` provides distinctive errors that are used to imply connection state.

andrewbaptist reviewed (changes requested):
I really like the general idea here, and want to do something similarly internally for determining which nodes are live (other than the liveness range). A few cleanups and this could be good!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-:-NUNAykD1GMME7gn8Dcw:b-cw6hta)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @koorosh, and @nvanbenschoten)

___
*[`pkg/gossip/gossip.go` line 1536 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN9q5N1mp-EQ_1g8Wq:-NUN9q5N1mp-EQ_1g8Wr:b-lswpnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/gossip/gossip.go#L1536)):*
> ```Go
> 	g.mu.RLock()
> 	defer g.mu.RUnlock()
> 	infos, err := g.mu.is.getMatchedInfos(MakePrefixPattern(KeyNodeDescPrefix))
> ```

There is already gossip.IterateInfos. Are you doing something different here so you can't use that? 
___
*[`pkg/server/status.go` line 1891 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN8buO70GT_eSLH0lU:-NUN8buO70GT_eSLH0lV:b-jwkezq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/status.go#L1891)):*
> ```Go
> }
> 
> // NetworkConnectivity returns information about connections statuses between nodes.
> ```

If this is going to initiate connections and not just report current statuses, this comment should change. I also was hoping to use something like this internally for a change I'm looking at to liveness. Maybe this could be separated into the needed parts (putting more of this in the rpc package) and only deal with "organizing" the data here.
___
*[`pkg/server/status.go` line 1904 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN839b6cbY-O9H-RBT:-NUN839c2Z5Zh6jsG-5B:b-fiyfp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/status.go#L1904)):*
> ```Go
> 
> 	if len(req.NodeID) > 0 {
> 		sourceNodeID, local, err := s.parseNodeID(req.NodeID)
> ```

You shouldn't need to parse here. Shouldn't the NodeId field be using a casttype to NodeID instead?
___
*[`pkg/server/status.go` line 1977 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN6Wmb4Z_K8RERMBhZ:-NUN6Wmb4Z_K8RERMBh_:byvss5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/status.go#L1977)):*
> ```Go
> 	}
> 
> 	if err := s.iterateNodes(ctx, "network connectivity", dialFn, nodeFn, responseFn, errorFn); err != nil {
> ```

It looks like this gets a list of all nodes and then attempts to dial all the ones it is not currently connected to. Today we mostly have a "full mesh" network where every node dials every other node, but this would "force" it to continue. I think it would be better to separate nodes into three classes:
1) Currently connected
2) Not connected (but not necessarily unable to connect).
3) Last connection attempt failed. 

I'm not sure how much harder that logic could be, but it seems incorrect to dial a node for reporting reason to find out if you can dial it. Is the intention that this would be called with every node in the system?
___
*[`pkg/server/serverpb/status.proto` line 1977 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN8Ns3CY12PVsbHkQI:-NUN8Ns3CY12PVsbHkQJ:bmu4ggf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/serverpb/status.proto#L1977)):*
> ```protobuf
> message NetworkConnectivityRequest {
>   string node_id = 1 [
>     (gogoproto.customname) = "NodeID"
> ```

You should add a casttype = "NodeID". I see you did that for the response, but is there a specific reason it doesn't work on the Request side?


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[`pkg/server/status.go` line 1977 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN6Wmb4Z_K8RERMBhZ:-NUVkzxO6n4hSz7bayJu:b-o9qfy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/status.go#L1977)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

It looks like this gets a list of all nodes and then attempts to dial all the ones it is not currently connected to. Today we mostly have a "full mesh" network where every node dials every other node, but this would "force" it to continue. I think it would be better to separate nodes into three classes:
1) Currently connected
2) Not connected (but not necessarily unable to connect).
3) Last connection attempt failed. 

I'm not sure how much harder that logic could be, but it seems incorrect to dial a node for reporting reason to find out if you can dial it. Is the intention that this would be called with every node in the system?

</blockquote></details>

I'd assume that this functionality should mostly rely on gossip to avoid calling other nodes. Similar functionality was removed from gossip recently 2114678be6d1ac279ec1fed3420fb635cc255476 and wondering if it could help.

``` go
// Connectivity returns the current view of the gossip network as seen by this
// node.
func (g *Gossip) Connectivity() Connectivity {
```
___
*[`pkg/server/serverpb/status.proto` line 1977 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN8Ns3CY12PVsbHkQI:-NUVmmxS-dqUnTapdbZn:bmti26x) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/serverpb/status.proto#L1977)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

You should add a casttype = "NodeID". I see you did that for the response, but is there a specific reason it doesn't work on the Request side?
</blockquote></details>

Request messages define NodeID as a plain string so clients (including Db Console) can either provide exact node ID or use special case, use "local" as node ID.
___
*[`pkg/gossip/gossip.go` line 1536 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN9q5N1mp-EQ_1g8Wq:-NUVkhRdAo82RS1nPGbK:bres6sr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/gossip/gossip.go#L1536)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

There is already gossip.IterateInfos. Are you doing something different here so you can't use that? 
</blockquote></details>

That exactly what I needed! Changed to use `gossip.IterateInfos`.
___
*[`pkg/ui/workspaces/db-console/src/util/api.ts` line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NM498aUYJ-N6jHMjDXi-r1-234:-NUVtyHN48XucooTTpFG:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f9ebf3537da999f1ed756afd9a09e7e74146caf/pkg/ui/workspaces/db-console/src/util/api.ts#L234)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Could you put all of the "front end" changes into a separate commit or PR?
</blockquote></details>

Done. 


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[`pkg/server/status.go` line 1977 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN6Wmb4Z_K8RERMBhZ:-NV-X7M3A1Uu6z7CWF9A:b-ot4uk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/status.go#L1977)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

I'd assume that this functionality should mostly rely on gossip to avoid calling other nodes. Similar functionality was removed from gossip recently 2114678be6d1ac279ec1fed3420fb635cc255476 and wondering if it could help.

``` go
// Connectivity returns the current view of the gossip network as seen by this
// node.
func (g *Gossip) Connectivity() Connectivity {
```
</blockquote></details>

Learned that gossip isn't a solution.


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[`pkg/server/status.go` line 1878 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NM498aUYJ-N6jHMjDXh-r1-1878:-NVTGEgQA01VEMyZudr5:b-cmjmji) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f9ebf3537da999f1ed756afd9a09e7e74146caf/pkg/server/status.go#L1878)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'm not sure we want to rely on the `bootstrapAddrs` field here. The semantics for that are a bit shaky.  Even though `bootstrapAddrs` seems like it very aggressively tracks remote addresses, I think we should try to build something that's more easily understood and that we can reason better about.

Each node gossips its NodeDescriptor here:

https://github.com/cockroachdb/cockroach/blob/2114678be6d1ac279ec1fed3420fb635cc255476/pkg/gossip/gossip.go#L403-L405

The TTL for that is 2h. So we can reasonably assume that a node will know all of its recent peers whenever network trouble begins. (And we cannot avoid this not always being true in contrived scenarios anyway: a node might restart while being partitioned, losing all memory state; yes we could add persistence but don't think we would as the persisted nodes could be stale after restart!)

You added an accessor for this this already: GetAllNodeIDs(). Now instead of listing just the IDs, it should also return the descriptors. Then you can construct a map from address to nodeID (don't forget about the LocalityAddrs which I just learned is a thing) and that can 

</blockquote></details>

1. Done. Address info is retrieved from node descriptors
2. > don't forget about the LocalityAddrs
that's interesting question. I didn't know about them either and would prefer to deal with this as a followup issue.
___
*[`pkg/server/status.go` line 1891 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN8buO70GT_eSLH0lU:-NVTFkyH0AHxPtJbHb9S:b-im2m80) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/status.go#L1891)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

If this is going to initiate connections and not just report current statuses, this comment should change. I also was hoping to use something like this internally for a change I'm looking at to liveness. Maybe this could be separated into the needed parts (putting more of this in the rpc package) and only deal with "organizing" the data here.
</blockquote></details>

Not sure if there's something that I could move to `rpc` package as far as it just aggregates things needed for Db Console.
___
*[`pkg/server/status.go` line 1904 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN839b6cbY-O9H-RBT:-NVTFpjR8leXSvRy97J2:bufbrtn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/status.go#L1904)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

You shouldn't need to parse here. Shouldn't the NodeId field be using a casttype to NodeID instead?
</blockquote></details>

`NodeID` in request payload is defined as a string to allow clients pass it as "local" (special case). It is common pattern for endpoints in Status Server.


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[`pkg/server/status.go` line 1977 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-NUN6Wmb4Z_K8RERMBhZ:-NV-rVq22fto76NIJ2IH:b-1nh3mo) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cafa7be22d94455dcc42d7f66a681d8b8dbcba9/pkg/server/status.go#L1977)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Learned that gossip isn't a solution.
</blockquote></details>

Agree, that looks incorrect however I don't see how it can be done efficiently with current architecture. Traversing nodes through established connections looks like more fundamental work on how nodes can communicate.


<!-- Sent from Reviewable.io -->

koorosh commented:
Updated PR to rely on changes from https://github.com/cockroachdb/cockroach/pull/99191 

Here's a quick loom video that shows how UI utilises this endpoint:
https://www.loom.com/share/919be1c8157e472d87546cdd701a968e?sid=518419a5-2238-4640-8863-c9699cda7255

tbg commented on pkg/server/status.go:
```diff
@@ -1989,6 +1990,104 @@ func (s *statusServer) NodeUI(
+				}
+				if err = s.rpcCtx.ConnHealth(node.Address.AddressField, node.NodeID, rpc.SystemClass); err != nil {
+					if kvpb.IsDecommissionedStatusErr(err) || errors.Is(rpc.ErrQuiescing, err) || grpcutil.IsClosedConnection(err) {
+						peer.Status = serverpb.NetworkConnectivityResponse_CLOSED
```
That doesn't seem quite right. We always close the connection when it enters an error state. It makes more sense to me to make this `NetworkconnectivityResponse_ERROR` and to also expose the error in the response (and show it in the UI tooltip).

tbg commented on pkg/server/status.go:
```diff
@@ -1989,6 +1990,104 @@ func (s *statusServer) NodeUI(
+					if kvpb.IsDecommissionedStatusErr(err) || errors.Is(rpc.ErrQuiescing, err) || grpcutil.IsClosedConnection(err) {
+						peer.Status = serverpb.NetworkConnectivityResponse_CLOSED
+					} else if errors.Is(rpc.ErrNotHeartbeated, err) {
+						peer.Status = serverpb.NetworkConnectivityResponse_NOT_CONNECTED
```
`ErrNotHeartbeated` means we are connecting for the first time. So I would call this `INITIALIZING` or something like that.

tbg commented on pkg/server/status.go:
```diff
@@ -1989,6 +1990,104 @@ func (s *statusServer) NodeUI(
+					} else if errors.Is(rpc.ErrNotHeartbeated, err) {
+						peer.Status = serverpb.NetworkConnectivityResponse_NOT_CONNECTED
+					} else {
+						peer.Status = serverpb.NetworkConnectivityResponse_PARTITIONED
```
This one would go away. I don't think we are in the right place here to reason about what a connection error "means"; the code should avoid trying to do that (it's not even something we can do in general). The connection is either absent, initializing, errored, or established. Exposing the error is more important than trying to classify it.

koorosh replied:
>  The connection is either absent, initializing, errored, or established.

@tbg , technically all connections are closed with errors, but would it make sense to indicate connections that were closed intentionally as errored? I'd try to distinguish those connections.

tbg replied:
I'm not sure, but it would also require another update to the `rpc` package. We can consider it, but let's not hold up this work for it.

If you are exposing the error text as I suggested elsewhere, the error message can also be a carried for this, so we may not need a new enum state. But we basically never close connections on purpose anyway.

koorosh replied:
Done. Updated connection statuses to rely on `ERROR` status and provide error info whenever this information is available.

tbg reviewed (commented):
First coarse review, the diff is a bit noisy due to the error rename, just pull that out (or even send a separate PR, then I can greenlight it immediately). My main comments are about what enum states to have.

koorosh reviewed (commented):
Ack. Will do. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95429#-:-NYX2urJAD-uCWb7rjz-:b-86qr7v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @Santamaura, and @tbg)

___
*[`pkg/server/status.go` line 2038 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95429#gh-1237010182:-NYX6_vg6Yl2QtrNyQFg:be2hewh) ([raw file](https://github.com/cockroachdb/cockroach/blob/42ad7a607e4cc03d92262feb6a8b806c33776c24/pkg/server/status.go#L2038)):*
I tried to distinguish two possible cases:
- connection closed as intended, planned shutdown, decommissioning, etc (so it's `_CLOSED` status)
- connection closed unexpectedly - that should bring attention of users. (set as `_PARTITIONED` status as all remaining errors)

Maybe it would be more clear to rename `_PARTITIONED` to `_ERROR`?



> expose the error in the response (and show it in the UI tooltip)

Correct, I'm going to show errors in tooltip for all "unhealthy" statuses.
___
*[`pkg/server/status.go` line 2040 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95429#gh-1237010949:-NYXB-Tm0mrHB_uZ2JkX:b-ti1lfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/42ad7a607e4cc03d92262feb6a8b806c33776c24/pkg/server/status.go#L2040)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`ErrNotHeartbeated` means we are connecting for the first time. So I would call this `INITIALIZING` or something like that.
</blockquote></details>

Agree. What about `_ESTABLISHING` (to make zero assumptions that connection is successful, just in process of establishing)
___
*[`pkg/server/status.go` line 2042 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95429#gh-1237013172:-NYXCKQ052lZ-RDVGuMT:b-34oyvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/42ad7a607e4cc03d92262feb6a8b806c33776c24/pkg/server/status.go#L2042)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This one would go away. I don't think we are in the right place here to reason about what a connection error "means"; the code should avoid trying to do that (it's not even something we can do in general). The connection is either absent, initializing, errored, or established. Exposing the error is more important than trying to classify it.
</blockquote></details>

Agree in general and as mentioned above, this case can be marked as `_ERROR` status without specific reasoning.
It is a case where we couldn't identify exact reason why connection closed (as it done with first case).
___
*[`pkg/rpc/context_test.go` line 2102 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95429#gh-1237006604:-NYXDURt3yzNHhkreBmS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/42ad7a607e4cc03d92262feb6a8b806c33776c24/pkg/rpc/context_test.go#L2102)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

please make a separate commit for the export of `errQuiescing`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg commented on pkg/server/status.go:
```diff
@@ -1989,6 +1989,104 @@ func (s *statusServer) NodeUI(
+		}
+
+		if local {
+			peers := map[roachpb.NodeID]serverpb.NetworkConnectivityResponse_Peer{}
```
// "local" specified, so collect the local results and return them back.

tbg commented on pkg/server/status.go:
```diff
@@ -1989,6 +1989,104 @@ func (s *statusServer) NodeUI(
+			}
+			return response, nil
+		}
+
```
// A non-local NodeID was specified, so relay to another node.

I think this control flow would be easier (and the bulk of code less indented) if you switched it to

```go
if len(req.NodeID) > 0 {
  sourceNodeID, local, err := s.parseNodeID(req.NodeID)
  if err != nil {
	  return nil, serverError(ctx, err)
  }}
  if !local {
    // do the stuff that's below this comment
  }
  // Do the stuff that is now in `if local`.
}
```

tbg commented on pkg/server/status_test.go:
```diff
@@ -1127,6 +1127,59 @@ func TestHotRanges2Response(t *testing.T) {
+			Knobs: base.TestingKnobs{
+				Store: &kvserver.StoreTestingKnobs{
+					ReplicaPlannerKnobs: plan.ReplicaPlannerTestingKnobs{
+						DisableReplicaRebalancing: true,
```
Do you want just this?

https://github.com/cockroachdb/cockroach/blob/42e45b4b8160f8a550a298b64f43799072fdfb5a/pkg/kv/kvserver/client_lease_test.go#L113-L119

The setup here seems bespoke. The snippet above will disable all queues, so all replicas will be only on n1, which seems appropriate for this test.

koorosh replied:
Thanks, that's what I needed!

tbg commented on pkg/server/status.go:
```diff
@@ -1989,6 +1989,104 @@ func (s *statusServer) NodeUI(
+				if err != nil {
+					peer.Status = serverpb.NetworkConnectivityResponse_UNKNOWN
+					cause := errors.UnwrapAll(err)
+					peer.Error = cause.Error()
```
`peer.Error = errors.UnwrapAll(err).Error()`

tbg commented on pkg/server/status.go:
```diff
@@ -1989,6 +1989,104 @@ func (s *statusServer) NodeUI(
+						peer.Status = serverpb.NetworkConnectivityResponse_ERROR
+					}
+					cause := errors.UnwrapAll(err)
+					peer.Error = cause.Error()
```
`peer.Error = errors.UnwrapAll(err).Error()`

tbg reviewed (approved):
Great! Nothing substantial left, only nits.

koorosh commented:
@andrewbaptist , I didn't address your suggestion regarding not using fan out of requests in this PR. Currently there's no easy way to avoid this.

andrewbaptist reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

koorosh commented:
bors retry
